feat(mobile-ui): Extend MetricsQueryBuilder and add gauge metrics

narsaynorath opened the pull request:
Extends the MetricsQueryBuilder to support gauges because we're trying to leverage gauges more and the mobile metrics for frame information are gauges.

I've added a `snql_gauge` method to the `MetricsFunction` class and make sure the function gets routed to the proper database table. The bulk of this PR is r.e. how I'm influencing the query builder to ensure that we set the right attributes to query the correct table, i.e. my `_evaluate_metric_matches_function` helper

narsaynorath committed:
feat(spans-metrics): Extend avg_if and avg_compare to frame metrics

Adds a SPAN_METRIC_COUNT_COLUMNS constant that exposes gauge metrics with
no unit. Also adds `frames_delay` to the SPAN_METRIC_DURATION_COLUMNS
definition by including seconds as the unit.

narsaynorath committed:
wip: Add metric to gauges

narsaynorath committed:
Allow gauges on MetricsQueryBuilder

narsaynorath committed:
Fix helper function call

narsaynorath committed:
Fix typing on store_span_metric

narsaynorath committed:
Cover more valid cases for function matching

narsaynorath commented on src/sentry/search/events/builder/metrics.py:
```diff
@@ -760,6 +778,29 @@ def _environment_filter_converter(self, search_filter: SearchFilter) -> WhereTyp
+        if type(expected_prefix) is str:
+            expected_prefix = {expected_prefix}
+
+        if metric_id is None or metric_id == -1:
+            return True
```
The `MetricQueryBuilderTest.test_multiple_references_only_resolve_index_once` test was providing `-1` as the `metric_id` for `measurements.lcp` and it would fail to resolve in the StringIndexer. Is this some sort of test artefact or does `-1` mean something for metric IDs, similar to how -1 means all projects in project selection?

shruthilayaj replied:
and I think if we do [this](https://github.com/getsentry/sentry/pull/69733/files#r1583644084), we don't have to handle that test case explicitly

narsaynorath replied:
Ah, I realized that the test was mocking the `-1` value so there isn't a code path that sets `-1` as the value for a metric ID. I'm going to get rid of this and mock the `reverse_resolve` for now. The test doesn't update the string indexer so the reverse resolve fails

shruthilayaj commented on src/sentry/search/events/constants.py:
```diff
@@ -352,7 +356,13 @@ class ThresholdDict(TypedDict):
     key
     for key, value in SPAN_METRICS_MAP.items()
-    if value.endswith("@millisecond") and value.startswith("d:")
+    if (value.endswith("@millisecond") and value.startswith("d:"))
```
do we need to do the `value.startswith(...)` check? can't we just look at the unit portion of the mri?

narsaynorath replied:
I don't think so. I kept them because I wanted to make my changes as isolated as possible, but I think it's totally fine to look at the unit portion here. Updated.

shruthilayaj commented on src/sentry/search/events/builder/metrics.py:
```diff
@@ -592,7 +593,11 @@ def resolve_snql_function(
         resolve_only: bool,
     ) -> SelectType | None:
-        if snql_function.snql_distribution is not None:
+        metric_id = arguments.get("metric_id")
```
does `arguments.get("column")` work here to get the string mri? this prevents us from having to query the indexer (ie reverse_resolve) in the `_evaluate_metric_matches_function` later

shruthilayaj replied:
so can we do something like: 
1. if we find a valid prefix (ie d, s, g, c) and if the corresponding `snql_*` function is not defined, return early with a helpful error message
2. 
if snql_function.snql_distribution is not None and (if prefix = d or prefix is None):
  resolved_function = blah
if snql_function.snql_set is not None and (if prefix = s or prefix is None):
  resolved_function = blah
  
 so this will maintain existing behaviour and maybe more readable?

narsaynorath replied:
It doesn't seem like `arguments.get("column")` returns the MRI, in my test case it's coming through as the column name `"mobile.slow_frames"`. The full payload for `arguments` looks like this:
```
{'column': 'mobile.slow_frames', 'if_col': 'release', 'if_val': 'foo', 'metric_id': 10016}
```

I guess this is because when we define the snql resolvers, we always use the `metric_id` to match for values and the other values are just the other args passed along. I could create a new `calculated_arg` to surface the MRI, but I don't think it's worth hoisting that logic up because it's used so deep in the query builder.

narsaynorath committed:
Only check for suffix

narsaynorath committed:
Better var naming

narsaynorath committed:
Remove -1 case

The test mocks in the -1 metric ID. Since it doesn't actually get added
to metrics, I also mock the `reverse_resolve` so we get a proper MRI as
expected

narsaynorath committed:
Scope prefix check to only spans

narsaynorath committed:
Add helper that avoids indexer

narsaynorath committed:
Only do the check for metrics or span metrics

shruthilayaj commented on src/sentry/search/events/builder/metrics.py:
```diff
@@ -1307,6 +1323,36 @@ def get_default_value(meta_type: str) -> Any:
+        }
+
+        prefix = None
+        primary_metric = metrics_map[self.use_case_id].get(column)
```
nit: let's do something like 
```suggestion
        primary_metric = metrics_map[self.use_case_id].get(column, column)
```
in case someone doesn't define a public alias for an mri

shruthilayaj commented on src/sentry/search/events/builder/metrics.py:
```diff
@@ -1307,6 +1323,36 @@ def get_default_value(meta_type: str) -> Any:
+        prefix = None
+        primary_metric = metrics_map[self.use_case_id].get(column)
+        prefix = primary_metric.split(":")[0] if primary_metric else None
+        if prefix is not None and prefix_to_function_map.get(prefix) is None:
```
nit: let's do so any unknown prefixes get default treatment?
```suggestion
        if prefix is not in prefix_to_function_map and prefix_to_function_map.get(prefix) is None:
```

shruthilayaj commented on src/sentry/search/events/builder/metrics.py:
```diff
@@ -1307,6 +1323,36 @@ def get_default_value(meta_type: str) -> Any:
+            UseCaseID.SPANS,
+            UseCaseID.TRANSACTIONS,
+        }:
+            return None
```
nit: should we also early return if column is None?

shruthilayaj commented on src/sentry/search/events/builder/metrics.py:
```diff
@@ -1307,6 +1323,36 @@ def get_default_value(meta_type: str) -> Any:
+        primary_metric = metrics_map[self.use_case_id].get(column)
+        prefix = primary_metric.split(":")[0] if primary_metric else None
+        if prefix is not None and prefix_to_function_map.get(prefix) is None:
+            raise IncompatibleMetricsQuery(
```
and maybe a test case for this assertion error?

narsaynorath committed:
Handle custom measurements

narsaynorath committed:
Add test for invalid configuration

narsaynorath committed:
Add early exit for columns that aren't metrics

narsaynorath committed:
Add early exit condition for no prefix defined

narsaynorath committed:
Fix mock
